radix-engine/src/kernel/call_frame.rs

```diff
@@ -557,11 +560,12 @@ pub enum SubstateDiffError {
     ContainsDuplicateOwns,
 }
 
-#[derive(Debug, Default)]
+#[derive(Debug)]
 pub struct CallFrameInit<C> {
     pub data: C,
     pub global_addresses: IndexSet<GlobalAddress>,
     pub direct_accesses: IndexSet<InternalAddress>,
+    pub always_visible_global_nodes: &'static IndexSet<NodeId>,
```

Comment:
I'm a little confused why this is on `CallFrameInit`.

I think in the future, possibly having a flag `uses_always_visible_global_nodes` might make sense here, but for now it'd always be true.

I feel that the `always_visible_global_nodes` is a global property, so should conceptually live in the kernel itself, probably? And then it can be provided when the call to `get_node_visibility` comes in:

```rust
    fn kernel_get_node_visibility(&self, node_id: &NodeId) -> NodeVisibility {
        self.stacks.cur().get_node_visibility(self.always_visible_references, node_id)
    }
```

Then we don't need to have always visible global nodes here, or have system care about it.

Reply:
> I think in the future, possibly having a flag uses_always_visible_global_nodes might make sense here, but for now it'd always be true.

It could be modeled as a flag. I feel like a set is better here than using a flag since it feels slightly nicer.

> I feel that the always_visible_global_nodes is a global property, so should conceptually live in the kernel itself, probably? And then it can be provided when the call to get_node_visibility comes in:

We could have the globally visible nodes be stored in the kernel and then passed to `get_node_visibility` inside of `kernel_get_node_visibility`. However, the primary issue I see with this is that `get_node_visibility` gets called from other places that are not `kernel_get_node_visibility` and that do not have access to kernel state such as `CallFrame::pass_message` and `CallFrame::process_input_substate_key`. This would mean that we still have to have the globally visible nodes be stored in the callframe or at least accessible to it.

Reply:
Ahh fair enough! Happy to keep this as-is then :)